test/stats_test.go

```diff
@@ -0,0 +1,122 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package test
+
+import (
+	"context"
+	"net"
+	"sync"
+	"testing"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/interop"
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/stats"
+)
+
+// TestPeerForClientStatsHandler tests the scenario where stats handler
+// (having peer as part of its struct) has peer enriched as part of
+// stream context.
+func (s) TestPeerForClientStatsHandler(t *testing.T) {
+	spy := &handlerSpy{}
+
+	// Start server.
+	l, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	grpcServer := grpc.NewServer()
+	testgrpc.RegisterTestServiceServer(grpcServer, interop.NewTestServer())
+	errCh := make(chan error)
+	go func() {
+		errCh <- grpcServer.Serve(l)
+	}()
+	t.Cleanup(func() {
+		grpcServer.Stop()
+		if err := <-errCh; err != nil {
+			t.Error(err)
+		}
+	})
+
+	// Create client with stats handler and do some calls.
+	conn, err := grpc.NewClient(
+		l.Addr().String(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithStatsHandler(spy))
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() {
+		if err := conn.Close(); err != nil {
+			t.Error(err)
+		}
+	})
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	client := testgrpc.NewTestServiceClient(conn)
+	interop.DoEmptyUnaryCall(ctx, client)
+	interop.DoLargeUnaryCall(ctx, client)
+	interop.DoClientStreaming(ctx, client)
+	interop.DoServerStreaming(ctx, client)
+	interop.DoPingPong(ctx, client)
+
+	// Assert if peer is populated for each stats type.
+	for _, callbackArgs := range spy.Args {
+		if callbackArgs.Peer == nil {
+			switch callbackArgs.RPCStats.(type) {
+			case *stats.Begin:
+				continue
+			case *stats.PickerUpdated:
+				continue
+			default:
+			}
+			t.Errorf("peer not populated for: %T", callbackArgs.RPCStats)
+		}
+	}
+}
```

Comment:
This is from the issue, but I don't think this test is actually verifying what we want. This only verifies that it never has received a peer call in it's callbackArgs. However, it does not verify that it actually received the peer, since this doesn't assert on the happy case it received a peer, just that a single callout didn't get a peer. What if I only got a Begin callout without the peer and the other callouts don't get peer, this test would pass. Maybe define how many callouts/peers in those callouts or something?

Reply:
```
for _, callbackArgs := range statsHandler.Args {
		if callbackArgs.peer == nil {
			switch callbackArgs.rpcStats.(type) {
			case *stats.Begin:
				continue
			case *stats.PickerUpdated:
				continue
			default:
			}
			t.Errorf("peer not populated for: %T", callbackArgs.rpcStats)
		}
	}
```

@zasweq Here we are making sure we received the peer for all the stats handlers except Begin and PickerUpdated. If we don't receive the peer for let's say OutPayload(and similary for every others except begin and pUpdated), it would go to switch:default and then `t.Errorf("peer not populated for: %T", callbackArgs.rpcStats)` and eventually fail.

Reply:
done

Reply:
Please switch to exact equality check.

Reply:
done